### PR TITLE
Implement the use of MNAIO images for system tests

### PR DESCRIPTION
**Pull request by odyssey4me**
This PR does the following:

1. To make testing the same way as the gate simpler outside of the gate, we implement a common vars script which is sourced by the pre/run/post stages, ensuring that all vars required for each stage are available.

2. Currently the tox tests are executed via multiple layers of scripts. This patch flattens it into the gating/check/run script to make following the code simpler.

3. We move the pre-installation of distro and python packages to the 'pre' stage to improve the chances of success, given that the 'pre' script is run three times before failing.

4. Currently the AIO deploy tests are executed via multiple layers of scripts. This patch flattens it into the gating/check/run script to make following the code simpler.

5. We move the MNAIO tempest run into run_deploy_mnaio.sh to reduce complexity and make the code easier to follow.

6. We cut down on duplicated var output between the 'prep' and 'run' stages.

7. Does all the MNAIO host preparation in the 'pre' stage of the gating pipeline. This stage is tried three times before failing, improving the chances of success. The actions done in this stage all relate to the VM preparation, and have nothing to do with deploying RPC-O.

8. Implements a check to verify whether there is an MNAIO image manifest available for this configuration. If so, it sets the preparation stage to download the images and use those to start the VM's.

9. If no images are available, or if this is a 'deploy' test, then build the VM's as usual.

10. Switches the 'system' tests to use the images if they are available.

Issue: [RE-1562](https://rpc-openstack.atlassian.net/browse/RE-1562)

### CODE REVIEW

**Comment by odyssey4me** [comment]
I'm using the MNAIO sdqc tests to validate the existing images. If you're getting failure reports, please ignore them - apologies for the noise. Test running here: https://rpc.jenkins.cit.rackspace.net/view/RPC-O/job/PM_rpc-openstack-master-xenial_mnaio_no_artifacts-swift-sdqc/110

**Comment by odyssey4me** [comment]
The ``sdqc`` jobs have been failing since Sep 13, (the last success being Sep 12)[https://rpc.jenkins.cit.rackspace.net/view/RPC-O/job/PM_rpc-openstack-master-xenial_mnaio_no_artifacts-swift-sdqc/98/]. The job result does show that the images were used though.

**Comment by odyssey4me** [comment]
I'm running a test of the ``deploy`` job. This patch should fix the current failures for the job due to the removal of the cryptography library install, which doesn't work for a bionic MNAIO host.

https://rpc.jenkins.cit.rackspace.net/view/RPC-O/job/PM_rpc-openstack-master-xenial_mnaio_no_artifacts-swift-deploy/319/

File: gating/check/pre [hunk 1]
```diff
@@ -24,3 +26,27 @@ fi
 # yaml library are present for ansible to work.
 source "$(readlink -f $(dirname ${0}))/../gating_prerequisites.sh"
 
+# Install pre-requisites for the different types of tests
```
Comment (prometheanfire) [inline gating/check/pre]: Bash arrays could probably be used here.

File: gating/check/pre [hunk 2]
```diff
@@ -24,3 +26,27 @@ fi
...
+distro_pkgs_to_install=""
+py_pkgs_to_install=""
+
+if [[ $RE_JOB_IMAGE =~ .*mnaio.* ]]; then
+  dpkg-query --list | grep curl &>/dev/null || distro_pkgs_to_install+="curl "
+fi
+
+if [[ $RE_JOB_ACTION == "tox-test" ]]; then
+  dpkg-query --list | grep wget &>/dev/null || distro_pkgs_to_install+="wget "
+  pip freeze | grep tox &>/dev/null || py_pkgs_to_install+="tox==3.3.0 "
```
Comment (prometheanfire) [inline gating/check/pre]: Can we use constraints here?

Comment (odyssey4me) [reply gating/check/pre]: We could. Are you thinking an in-repo set of constraints?

File: gating/check/pre (hunk 1 above)
Comment (odyssey4me) [reply gating/check/pre]: Sure, they could. This works, and is pretty simple though. If you feel strongly about it I can change it up.

Comment (prometheanfire) [reply gating/check/pre]: ya, this was a VERY minor nit

File: gating/check/pre (hunk 2 above)
Comment (prometheanfire) [reply gating/check/pre]: I was thinking pointing upstream somewhere so we don't have another thing to maintain (within the file here, or our repo).

Comment (odyssey4me) [reply gating/check/pre]: Well, pointing to a web address might be nice, but it does introduce another opportunity for failure. We could easily use the dep_update job to prepare the constraints file and propose changes along with all other dependencies. It should be a small set. I don't see tox in https://github.com/openstack/requirements/blob/master/upper-constraints.txt anyway.

**Comment by odyssey4me** [comment]
Fixed a bug, running a fresh test now: https://rpc.jenkins.cit.rackspace.net/view/RPC-O/job/PM_rpc-openstack-master-xenial_mnaio_no_artifacts-swift-deploy/320/

**Comment by odyssey4me** [comment]
recheck_dep_update

File: gating/check/pre (hunk 2 above)
Comment (prometheanfire) [reply gating/check/pre]: Could work (dep_update), also funny that we don't have tox I don't see anything preventing us from adding it.

I'd like to see reuse of the requirements code for managing constraints if possible (that's one of the things I'm working on upstream).

Comment (odyssey4me) [reply gating/check/pre]: Perhaps it'd be best to create a new card to implement this then, and leave this as-is? It's quite a divergence from the focus of my current card. For the tox install, all I did was move it and set a pin here - previously there was no pin.

Comment (prometheanfire) [reply gating/check/pre]: As long as we are tracking it, sure.

Comment (odyssey4me) [reply gating/check/pre]: I'm suggesting that you register a card in RI given that it's something you'd like to see as an improvement and have the domain knowledge for it. :)

Comment (odyssey4me) [reply gating/check/pre]: I'd be happy to help out, of course.

Comment (prometheanfire) [reply gating/check/pre]: lol, I knew it'd come down to this :P

Comment (odyssey4me) [reply gating/check/pre]: For reference, the card is here: https://rpc-openstack.atlassian.net/browse/RI-475

**Comment by odyssey4me** [comment]
Testing this patch combined with https://github.com/rcbops/rpc-gating/pull/1086 here: https://rpc.jenkins.cit.rackspace.net/view/RPC-O/job/PM_rpc-openstack-master-xenial_mnaio_no_artifacts-swift-deploy/322/

**Pull request merged**